driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -652,129 +561,135 @@ public ByteBuf getBuffer(final int size) {
         return stream.getBuffer(size);
     }
 
-    private class ResponseHeaderCallback implements SingleResultCallback<ByteBuf> {
+    private static <T extends BsonDocument> T getResponseDocument(final ResponseBuffers responseBuffers,
+                                                                  final int messageId, final Decoder<T> decoder) {
+        ReplyMessage<T> replyMessage = new ReplyMessage<T>(responseBuffers, decoder, messageId);
+        responseBuffers.reset();
+        return replyMessage.getDocuments().get(0);
+    }
+
+    private class MessageHeaderCallback implements SingleResultCallback<ByteBuf> {
         private final SingleResultCallback<ResponseBuffers> callback;
 
-        ResponseHeaderCallback(final SingleResultCallback<ResponseBuffers> callback) {
+        MessageHeaderCallback(final SingleResultCallback<ResponseBuffers> callback) {
             this.callback = callback;
         }
 
         @Override
-        public void onResult(final ByteBuf result, final Throwable throwableFromCallback) {
-            if (throwableFromCallback != null) {
-                callback.onResult(null, throwableFromCallback);
-            } else {
-                try {
-                    ReplyHeader replyHeader;
-                    ByteBufferBsonInput headerInputBuffer = new ByteBufferBsonInput(result);
-                    try {
-                        MessageHeader messageHeader = new MessageHeader(headerInputBuffer, description.getMaxMessageSize());
-                        replyHeader = new ReplyHeader(headerInputBuffer, messageHeader);
-                    } finally {
-                        headerInputBuffer.close();
-                    }
-
-                    if (replyHeader.getMessageLength() == TOTAL_REPLY_HEADER_LENGTH) {
-                        onSuccess(new ResponseBuffers(replyHeader, null));
-                    } else {
-                        readAsync(replyHeader.getMessageLength() - TOTAL_REPLY_HEADER_LENGTH,
-                                  new ResponseBodyCallback(replyHeader));
-                    }
-                } catch (Throwable t) {
-                    close();
-                    callback.onResult(null, t);
-                }
-            }
-        }
-
-        private void onSuccess(final ResponseBuffers responseBuffers) {
-
-            if (responseBuffers == null) {
-                callback.onResult(null, new MongoException("Unexpected empty response buffers"));
-                return;
-            }
-
+        public void onResult(final ByteBuf result, final Throwable t) {
             try {
-                callback.onResult(responseBuffers, null);
-            } catch (Throwable t) {
-                LOGGER.warn("Exception calling callback", t);
+                if (t != null) {
+                    throw t;
```

Comment:
Just use the callback directly, I don't like the throw and catch control flow, especially as it can be avoided here.